### PR TITLE
wasm httpCall & grpcCall support activeSpan

### PR DESCRIPTION
**Pull request by StarryVae**
Signed-off-by: wangkai19 <wangkai19@corp.netease.com>

Commit Message: wasm httpCall & grpcCall support activeSpan
Additional Description: N/A
Risk Level: low
Testing: N/A
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
issue #22028 cc @PiotrSikora

**Review by PiotrSikora — COMMENTED** [review]
Looks good, but could you add some tests? Otherwise, it's hard to tell if this actually changes anything.

**Comment by StarryVae** [comment]
> Looks good, but could you add some tests? Otherwise, it's hard to tell if this actually changes anything.

thanks, and i have some questions about the wasm tests, i have no idea why `decoder_callbacks_` is always `null` in `AsyncCall` and `GrpcCall` test. Ask for help @PiotrSikora , thanks.
https://github.com/envoyproxy/envoy/blob/46cfdb57d552a58a0ad109ef313d004eb4eb7adc/test/extensions/filters/http/wasm/wasm_filter_test.cc#L990

**Comment by StarryVae** [comment]
@tmsnan test

File: source/extensions/common/wasm/context.cc [hunk 1]
```diff
@@ -970,6 +970,9 @@ WasmResult Context::httpCall(std::string_view cluster, const Pairs& request_head
   Protobuf::RepeatedPtrField<HashPolicy> hash_policy;
   hash_policy.Add()->mutable_header()->set_header_name(Http::Headers::get().Host.get());
   options.setHashPolicy(hash_policy);
+  if (decoder_callbacks_) {
+    options.setParentSpan(decoder_callbacks_->activeSpan());
+  }
```
[the remark this answers is not in the document]
Comment (StarryVae) [reply source/extensions/common/wasm/context.cc:975]: `Http::AsyncClient::StreamOptions` do not have set span api and some tests are added, PTAL again, thanks. @PiotrSikora

[the remark this answers is not in the document]
Comment (PiotrSikora) [reply source/extensions/common/wasm/context.cc:975]: I believe that `Http::AsyncClient::StreamOptions` could be simply changed to `Http::AsyncClient::RequestOptions`.

**Review by PiotrSikora — COMMENTED** [review]
Regarding tests, you should verify that those spans are emitted as HTTP headers when tracers are configured.

**Comment by StarryVae** [comment]
> Regarding tests, you should verify that those spans are emitted as HTTP headers when tracers are configured.

@PiotrSikora thanks, some tests are added, and i think span is not supported in `GoogleAsyncStreamImpl` since it will not use `parent_span` in `RequestOptions`.

File: test/extensions/filters/http/wasm/wasm_filter_test.cc [hunk 2]
```diff
@@ -716,18 +750,31 @@ TEST_P(WasmHttpFilterTest, AsyncCall) {
...
                                                       {":authority", "foo"},
                                                       {"content-length", "11"}}),
                       message->headers());
             EXPECT_EQ((Http::TestRequestTrailerMapImpl{{"trail", "cow"}}), *message->trailers());
             callbacks = &cb;
+            current_span =
+                options.parent_span_->spawnChild(Tracing::EgressConfig::get(), "fake", start_time_);
+            current_span->injectContext(message->headers(), nullptr);
+            EXPECT_TRUE(message->headers().getByKey("x-datadog-parent-id") != "");
+            EXPECT_TRUE(message->headers().getByKey("x-datadog-parent-id") != "");
```
Comment (PiotrSikora) [inline test/extensions/filters/http/wasm/wasm_filter_test.cc:776]: Should one of those be `x-datadog-trace-id` instead?

File: source/extensions/common/wasm/context.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply source/extensions/common/wasm/context.cc:975]: Did you try doing it?

File: test/extensions/filters/http/wasm/wasm_filter_test.cc (hunk 2 above)
Comment (StarryVae) [reply test/extensions/filters/http/wasm/wasm_filter_test.cc:776]: thanks, i will fix this

**Comment by StarryVae** [comment]
@PiotrSikora it seems that datadog trace test is not supported in windows.

**Comment by PiotrSikora** [comment]
> @PiotrSikora it seems that datadog trace test is not supported in windows, should i add `skip_on_windows` in BUILD file?

Are any other tracers supported on Windows?

**Comment by StarryVae** [comment]
> > @PiotrSikora it seems that datadog trace test is not supported in windows, should i add `skip_on_windows` in BUILD file?
> 
> Are any other tracers supported on Windows?

i will take a look

**Comment by StarryVae** [comment]
@PiotrSikora i changed to the `zipkin tracing`, PTAL again, thanks.

**Review by PiotrSikora — COMMENTED** [review]
Perhaps I misconfigured something, but I don't see trace/span IDs in HTTP request headers for HTTP/gRPC calls, and I don't see thoes calls in Zipkin UI.

Did you verify that this works in production code and not only in tests? Could you share a sample working configuration?

**Comment by StarryVae** [comment]
> Perhaps I misconfigured something, but I don't see trace/span IDs in HTTP request headers for HTTP/gRPC calls, and I don't see thoes calls in Zipkin UI.
> 
> Did you verify that this works in production code and not only in tests? Could you share a sample working configuration?

@PiotrSikora i tried this using examples/wasm-cc/envoy_filter_http_wasm_example.cc, and i have no idea why `decoder_callbacks_` of wasm context is always `NULL,` which results `options.setParentSpan` will never be called.

Here is part of code i add to the `onRequestHeaders` in examples/wasm-cc/envoy_filter_http_wasm_example.cc:

`  auto success = root()->httpCall("web_service", headers, "", trailers, 5000,
                                 [](uint32_t, size_t, uint32_t) {});`

**Comment by PiotrSikora** [comment]
> @PiotrSikora i tried this using examples/wasm-cc/envoy_filter_http_wasm_example.cc, and i have no idea why `decoder_callbacks_` of wasm context is always `NULL,` which results `options.setParentSpan` will never be called.
> 
> Here is part of code i add to the `onRequestHeaders` in examples/wasm-cc/envoy_filter_http_wasm_example.cc:
> 
> ` auto success = root()->httpCall("web_service", headers, "", trailers, 5000, [](uint32_t, size_t, uint32_t) {});`

Hmm... The current implementation dispatches HTTP/gRPC callouts from the root context (i.e. a plugin context, and not a  context associated with a particular HTTP request) in order to allow calls from background tasks, timers, etc., and to protect against various use-after-free scenarious in case HTTP request completes before response from HTTP/gRPC callout is processed.

I don't see a reason why it has to stay this way, and we should be able to associate HTTP/gRPC callouts with HTTP request if they're dispatched from any HTTP callback, but it's a bigger change than this PR alone.

cc @mpwarres

**Comment by alyssawilk** [comment]
@PiotrSikora @mpwarres not sure of the status of this one - should it be tagged as waiting?

**Comment by PiotrSikora** [comment]
> @PiotrSikora @mpwarres not sure of the status of this one - should it be tagged as waiting?

Yes, thanks!

**Comment by jmarantz** [comment]
/wait

looks like this needs a main merge.

**Comment by tkennedy1-godaddy** [comment]
Did this get closed by mistake? Or is this PR willing to be re-opened? We'd be happy to start pitching in on this!

**Comment by jmarantz** [comment]
If the original author can't push it forward, maybe you can patch it into a new PR?

**Comment by StarryVae** [comment]
> If the original author can't push it forward, maybe you can patch it into a new PR?

sorry for late, i will pick this up again.

**Comment by StarryVae** [comment]
> examples/wasm-cc/envoy_filter_http_wasm_example.cc

@PiotrSikora Hi, sorry for late reply, could you show me how to dispatche HTTP/gRPC callouts from the root context in the `examples/wasm-cc/envoy_filter_http_wasm_example.cc`, thanks.

**Comment by juanmolle** [comment]
Was this PR closed by mistake?

**Pull request closed without merging**